### PR TITLE
Fix promises

### PR DESCRIPTION
**Pull request by kirill-gerasimenko**
Hi,

This is a part of the PR I'm working on atm for ionide-vscode-fsharp, and since it is using ionide-vscode-helpers I would need to have these changes here first.

I started to work on adding setting for Mono path and noticed that I couldn't chain promises the way I expect. And seems like I've found the issue with they way some Promises module's method work. 
The issue is that in Promise.success the onrejected handler is a function returning unit, which is a "catch" handler, meaning that if you return value from it, like unit, int, etc. it is considered as handled and returned promise is not in rejected state. So when chaining promises with |> operator if some handler fails - stops at first Promise.fail handler and it returns a new promise with "pending" state. I hope it makes sense.

So after playing with the current API I made some changes, along with "fixing" it to conform with behavior I expect. 

Here is the list of methods:
- Promise.create - handler is provided with "resolve" and "reject" methods allowing some async scenarios
- Promise.map - previously was "success", I've decided that it reflects the purpose better, and we have "bind" already. 
- Promise.bind - remained the same, just fixed the implementation for better chainability.
- Promise.catch - made some changes to implementation there using "then", instead of "catch", which made the example below to compile properly. Doing Promise.then(null, handler) is semantically the same as "catch". 
- Promise.reject<'T> - added this one (previously added it to Fable), this allows for example return (Promise.reject<ChildProcess> None) where you need to reject the promise, and typed version works better, than just Promise.reject<unit> - it compiles in situations like this: 

``` fsharp
let p = promise { 
   <...>
   return!
       if <...> then 
           Promise.lift 10
       else
           Promise.reject<int> None
}
```
- Promise.onSuccess - this is purely for side effects (like logging, or just some actions to be done), return result is unit, the value is propagated to the returned promise
- Promise.onFail - this is purely for side effects, same as onSuccess but for rejected promise. Propagates error value - better for chaining.  

Here is test example for the successful part of things: 

``` fsharp
Promise.create (Func<_,_,_>(fun resolve reject -> 
    resolve.Invoke (U2.Case1 "Succeeded promise")
))
|> Promise.onFail (fun error -> printfn "Error1: %A" error |> ignore)
|> Promise.onFail (fun error -> printfn "Error2: %A" error |> ignore)
|> Promise.onSuccess (fun value -> printfn "Value: %A" value)
|> Promise.map (fun value -> "Mapped value")
|> Promise.onSuccess (fun value -> printfn "Value after map: %A" value)
|> Promise.bind (fun value -> Promise.lift 1.5)
|> Promise.onSuccess (fun value -> printfn "Value after Promise.bind: %A" value)
|> Promise.onFail (fun error -> printfn "Error3: %A" error)
|> Promise.catch (fun error -> promise { 
    printfn "Catched value: %A" |> ignore
    return 100
    })
|> Promise.onSuccess (fun intValue -> printf "INT VALUE IS: %d" intValue)
|> ignore
```

When executed it gives this result: 

```
Value: "Succeeded promise"
Value after map: "Mapped value"
Value after Promise.bind: 1.5
INT VALUE IS: 1.5
```

When instead of `resolve.Invoke (U2.Case1 "Succeeded promise")` we call `reject.Invoke "Failed promise"` we get this: 

```
Error1: "Failed promise"
Error2: "Failed promise"
Error3: "Failed promise"
Catched value: "Failed promise"
INT VALUE IS: 100
```

In ionide-fsharp I'm preparing another PR with custom Mono path setting and the Promises version from this PR and some related promises refactoring (for example LanguageService.start() is using lambda for continuation, but this in my opinion kind of breaks the promises idea there, and changes in PR are to fix these issues). 

I hope it all makes sense, this is pretty much my first attempts to do something useful here - so don't judge me very strong please :)

PS: link to [ionide-vscode-fsharp PR](https://github.com/ionide/ionide-vscode-fsharp/pull/134)

### CODE REVIEW

File: Helpers.fs [hunk 1]
```diff
@@ -16,29 +16,45 @@ module Promise =
     open Fable.Core
     open Fable.Import
     open Fable.Import.JS
+    
+    let create resolver = Promise.Create resolver
```
Comment (adamchester) [inline Helpers.fs]: Could we remove the `Func<>` issues by doing it this way?

``` fsharp
let create (body: ('T->unit) -> (obj->unit) -> unit) =
    Promise.Create(fun (resolverFunc : Func<U2<_, _>, _>) (rejectorFunc : Func<_,_>) ->
        body (fun v -> resolverFunc.Invoke(U2.Case1 v)) (fun e -> rejectorFunc.Invoke e))
```

**Comment by cloudRoutine** [comment]
@kirill-gerasimenko  Add some documentation comments and it's good to merge :+1:

**Comment by cloudRoutine** [comment]
Although if @adamchester's suggestion is viable, I would prefer that implementation

File: Helpers.fs (hunk 1 above)
Comment (kirill-gerasimenko) [inline Helpers.fs]: Yes, that looks much cleaner! I'll update the branch

**Comment by kirill-gerasimenko** [comment]
I'll add suggestion from @adamchester and documentation comments, thanks!

File: Helpers.fs [hunk 2]
```diff
@@ -40,18 +57,32 @@ module Promise =
     let either a  (b: Func<obj, U2<'R, PromiseLike<'R>>>) (pr : Promise<'T>) : Promise<'R> =
```
Comment (cloudRoutine) [inline Helpers.fs]: Can this `Func` be shifted into the body of `either` so that it takes a normal F# function as its arg instead?

Comment (kirill-gerasimenko) [inline Helpers.fs]: Do you mean make it the same way as @adamchester suggested for Promise.create?

Comment (cloudRoutine) [inline Helpers.fs]: yea

Comment (kirill-gerasimenko) [inline Helpers.fs]: Fixed :)

**Comment by cloudRoutine** [comment]
LGTM

**Comment by kirill-gerasimenko** [comment]
Great, thanks a lot!

**Pull request merged**